python/src/equistore/operations/join.py

```diff
@@ -0,0 +1,212 @@
+                ]
+            ).T
+
+        property_names = ("tensor", "property")
+    else:
+        # if property_names are the same in each block and the values are
+        # unique we keep the names of the first block
+        property_names = names1
+
+    properties_values = np.vstack([properties1, properties2])
+
+    return Labels(names=property_names, values=properties_values)
+
+
+def _join_blocks_along_properties(
+    block1: TensorBlock, block2: TensorBlock
+) -> TensorBlock:
+    if not np.all(block1.samples == block2.samples):
+        raise ValueError("`A` and `B` do not have the same samples.")
+
+    if block1.components != block2.components:
+        raise ValueError("`A` and `B` do not have the same components")
+
+    properties = _join_labels(block1.properties, block2.properties)
+    result_block = TensorBlock(
+        values=_dispatch.hstack([block1.values, block2.values]),
+        samples=block2.samples,
+        components=[],
+        properties=properties,
+    )
+
+    for parameter, gradient1 in block1.gradients():
+        gradient2 = block2.gradient(parameter)
+
+        gradient_data = np.zeros(gradient1.data.shape[:-1] + (len(properties),))
+
+        # copy block1 gradient data over
+        gradient_data[..., : len(block1.properties)] = gradient1.data
+
+        # check where to put the block2 data
+        for i_grad, sample_grad in enumerate(gradient2.samples):
+            i_merged = gradient1.samples.position(sample_grad)
+            gradient_data[i_merged, ..., len(block1.properties) :] = gradient2.data[
+                i_grad
+            ]
+
+        result_block.add_gradient(
+            parameter, gradient_data, gradient1.samples, gradient1.components
+        )
+
+    return result_block
+
+
+def _join_blocks_along_samples(block1: TensorBlock, block2: TensorBlock) -> TensorBlock:
+    for property1, property2 in zip(block1.properties, block2.properties):
+        if tuple(property1) != tuple(property2):
+            raise ValueError("`A` and `B` do not have the same properties.")
+
+    if block1.components != block2.components:
+        raise ValueError("`A` and `B` do not have the same components")
+
+    result_block = TensorBlock(
+        values=_dispatch.vstack([block1.values, block2.values]),
+        samples=_join_labels(block1.samples, block2.samples),
+        components=[],
+        properties=block1.properties,
+    )
+
+    for parameter, gradient1 in block1.gradients():
+        gradient2 = block2.gradient(parameter)
+
+        n_samples_gradient = len(gradient1.samples) + len(gradient2.samples)
+        gradient_data = np.zeros((n_samples_gradient,) + gradient1.data.shape[1:])
+
+        gradient_data[: len(gradient1.samples)] = gradient1.data
+        gradient_data[len(gradient1.samples) :] = gradient2.data
+
+        result_block.add_gradient(
+            parameter,
+            gradient_data,
+            _join_labels(gradient1.samples, gradient2.samples),
```

Comment:
This is not the right way to join gradient samples. They should be concatenated vertically, and then the `sample` column should be updated to reflect the changes in the joined **values** samples.

Reply:
True I change it back to the previous version.